File: a2/python/sender.py
import sys
import threading
import time
from socket import *
from packet import packet

WINDOW_SIZE = 10
MAX_DATA_LENGTH = 500

class Sender:
  def __init__(self, emu_addr, emu_port, sender_port, file_name):
    super(Sender, self).__init__()
    self.emu_addr = emu_addr
    self.emu_port = emu_port
    self.sender_port = sender_port
    self.base = 0
    self.next_seq_num = 0
    self.timer = None

  def receive_ack(self, client_socket, file_ack):
    global WINDOW_SIZE

    while True:
      data, clientAddress = client_socket.recvfrom( 2048 )
      p = packet.parse_udp_data(data)
      file_ack.write(str(p.seq_num) + "\n")

      prev_base = self.base
      base_mod = self.base % packet.SEQ_NUM_MODULO
      cur_seq_num = p.seq_num
      windowEnd = (base_mod + WINDOW_SIZE - 1) % packet.SEQ_NUM_MODULO

      if p.type == 2:
        client_socket.close()
        break

      # check if within valid range
      if cur_seq_num >= base_mod and cur_seq_num <= windowEnd:
        self.base = self.base + cur_seq_num - base_mod + 1
      elif base_mod > windowEnd and (cur_seq_num <= windowEnd or cur_seq_num >= base_mod):
        if base_mod <= cur_seq_num:
          self.base = self.base + cur_seq_num + 1 - base_mod
        else:
          self.base = self.base + cur_seq_num + 1 + (packet.SEQ_NUM_MODULO - base_mod)
      
      if self.base > prev_base:
        if self.base != self.next_seq_num:
          self.timer = time.time()
        else:
          self.timer = None

  def start(self, file_name):
    # UDP Socket
    client_socket = socket(AF_INET, SOCK_DGRAM)
    client_socket.bind(('', self.sender_port))

    # generate three log files
    file_seqnum = open("seqnum.log","w+")
    file_ack = open("ack.log","w+")

    # read data from file and store it
    buffer = []
    with open(file_name, "rb") as f:
      byte = f.read(500)
      while byte:
        buffer.append(byte)
        byte = f.read(500)

    # start Acknowledger
    acknowledger = threading.Thread(target=self.receive_ack, args=(client_socket,file_ack,))
    acknowledger.start()

    # start sending packets
    total_packets = len(buffer)
    while not (self.base == total_packets):
      if self.next_seq_num < total_packets \
        and self.next_seq_num < self.base + WINDOW_SIZE:
        packet_to_send = packet.create_packet(self.next_seq_num, buffer[self.next_seq_num])
        
        client_socket.sendto(packet_to_send.get_udp_data(), (self.emu_addr, self.emu_port))
        file_seqnum.write(str(self.next_seq_num) + "\n")

        self.next_seq_num = self.next_seq_num + 1

        if self.base == self.next_seq_num:
          self.timer = time.time()
      elif timer and (time.time() - timer) > 0.1:
        # resend packets
        for i in range(self.base, self.next_seq_num):
          packet_to_send = packet.create_packet(i, buffer[i])

          client_socket.sendto(packet_to_send.get_udp_data(), (self.emu_addr, self.emu_port))
          file_seqnum.write(str(i) + "\n")

        timer = time.time()
    
    # send eot
    client_socket.sendto(packet.create_eot(-1).get_udp_data(), (self.emu_addr, self.emu_port))
    acknowledger.join()

    file_seqnum.close()
    file_ack.close()
    

############# MAIN STARTS HERE #############

def check_args(args):
  if len(args) != 4:
    raise Exception

  try:
    emu_addr = args[0]
    emu_port = int(args[1])
    sender_port = int(args[2])
    file_name = args[3]
  except:
    raise Exception

  return emu_addr, emu_port, sender_port, file_name

def main(args):
  try:
    emu_addr, emu_port, sender_port, file_name = check_args(args)
  except Exception:
    exit(-1)

  sender = Sender(emu_addr, emu_port, sender_port, file_name)

  # start sender class
  sender.start(file_name)

  print("Sender Finished!")

# Parameters:
#   $1 - host address of the network emulator
#   $2 - UDP port number used by the emulator to receive data from the sender
#   $3 - UDP port number used by the sender to receive ACKs from the emulator
#   $4 - name of the file to be transferred
main(sys.argv[1:])

File: a2/src/sender.java
import java.io.*;
import java.net.*;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.ReentrantLock;

public class sender {
	private static final int ACK = 0;
	private static final int DATA = 1;
	private static final int EOT = 2;
    private final static int windowSize = 10;
    private final static int maxDataLength = 500;
    private final static int timeoutLength = 100;

    private static InetAddress emulatorAddr;
    private static int emulatorPort;
    private static int senderPort;
    private static String fileName;

    private List<packet> packets;

    private PrintWriter seqnumLog, ackLog, timeLog;

    private ReentrantLock lock;

    private DatagramSocket socket;

    private long tStart;
    private long tEnd;

    private long baseTime;
    private boolean timerRunning = false;

    private int packetsSent = -1;
    private int packetsAcked = -1;
    private boolean packetSendingStarted = false;

    public static void main(String[] args) throws Exception{
        sender s = new sender();

        // init fields
        s.init(args);

        // init and start threads
        Thread t1 = new Thread(s.new AckReceiver());
        t1.start();
        
        Thread t2 = new Thread(s.new PacketSender());
        t2.start();

        // wait for threads to finish
        try {
            t1.join();
            t2.join();
        } catch (Exception e) {
            System.err.println("Sender: waiting thread crashed!");
        }

        // close files and sockets
        s.close();
    }

    private void init(String[] args) throws Exception {
        if (args.length != 4) {
            System.err.println("Sender: invalid number of arguments");
            System.exit(-1);
        }
        
        emulatorAddr = InetAddress.getByName(args[0]);
        try {
            // validate port numbers
            emulatorPort = Integer.parseInt(args[1]);
            senderPort = Integer.parseInt(args[2]);
        } catch (NumberFormatException e) {
            System.err.println("Sender: invalid port number given");
            System.exit(-1);
        }
        fileName = args[3];
        
        // create socket
        try {
            socket = new DatagramSocket(senderPort);
        } catch (SocketException e) {
            System.err.println("Sender: given port is not available");
        }

        // create files
        try {
            seqnumLog = new PrintWriter(new BufferedWriter(new FileWriter("seqnum.log")));
            ackLog = new PrintWriter(new BufferedWriter(new FileWriter("ack.log")));
            timeLog = new PrintWriter(new BufferedWriter(new FileWriter("time.log")));
        } catch (IOException e) {
            System.err.println("Sender: failed to create files");
        }

        // make packets
        try {
            packets = makePackets();
        } catch (Exception e) {
            System.err.println("Sender: failed to create packets");
        }

        // create lock
        lock = new ReentrantLock();
    }

    private void close() throws Exception {
        // close files
        try {
            seqnumLog.close();
            ackLog.close();
            timeLog.close();
        } catch (Exception e) {
            System.err.println("Sender: cannot close files");
        }

        // close socket
        try {
            socket.close();
        } catch (Exception e) {
            System.err.println("Sender: failed to close socket");
        }
    }

    private ArrayList<packet> makePackets() throws Exception {        
        ArrayList<packet> newPackets = new ArrayList<>();
        FileInputStream fis = new FileInputStream(fileName);
        
        int seqNum = 0;
        while(fis.available() > 0) {
            byte[] bytes = new byte[Math.min(maxDataLength, fis.available())];
            fis.read(bytes);
            String data = new String(bytes);

            newPackets.add(packet.createPacket(seqNum, data));
            seqNum++;
        }
        fis.close();

        return newPackets;
    }

    public void send(packet p) throws Exception {
        byte data[] = p.getUDPdata();
        DatagramPacket sendPacket = new DatagramPacket(data, data.length, emulatorAddr, emulatorPort);
        socket.send(sendPacket);
    }

    public void startTimer() throws Exception{
        baseTime = System.currentTimeMillis();
    }

    public void stopTimer() throws Exception {
        timerRunning = false;
    }

    /***********************************
            Packet Sender Thread
    ************************************/
    public class PacketSender implements Runnable {
        public void run() {
            try {
                sendPackets();
            } catch(Exception e) {
                e.printStackTrace();
            }
        }

        private void sendPackets() throws Exception {
            // send all packets
            while (packetsAcked < packets.size() - 1) {
                int base = packetsSent + 1;
                int windowRemaining = windowSize - (packetsSent - packetsAcked);

                if (windowRemaining > 0 && (packets.size() - 1) - packetsSent > 0) {
                    int offset = Math.min(windowRemaining, (packets.size() - 1) - packetsSent);
                    if (!timerRunning) {
                        lock.lock();
                            startTimer();
                            timerRunning = true;
                        lock.unlock();
                    }

                    for (int i = base; i < base + offset; i++) {
                        if (base == 0 && !packetSendingStarted) {
                            tStart = System.currentTimeMillis();
                        }
                        packetSendingStarted = true;

                        send(packets.get(i));
                        seqnumLog.println(packets.get(i).getSeqNum());
                    }

                    lock.lock();
                        packetsSent = base + offset - 1;
                    lock.unlock();
                }

                Thread.sleep(30);
                long deltaTime = System.currentTimeMillis() - baseTime;

                if (deltaTime >= timeoutLength && timerRunning) {
                    lock.lock();
                        stopTimer();
                        packetsSent = packetsAcked;
                    lock.unlock();
                }
            }

            // send EOT packet
            packet eotPacket = packet.createEOT(packets.size());
            send(eotPacket);
        }
    }

    /***********************************
            Ack Receiver Thread
    ************************************/
    public class AckReceiver implements Runnable{
        public void run() {
            try{
                receivePackets();
            } catch(Exception e) {
                e.printStackTrace();
            }
        }

        private void receivePackets() throws Exception {
            byte[] data = new byte[512];
            DatagramPacket receivePacket = new DatagramPacket(data, data.length);

            Thread.sleep(50);

            while(true) {
                // receive ACKed packets
                socket.receive(receivePacket);
                packet dataPacket = packet.parseUDPdata(data);

                if (dataPacket.getType() == EOT) break;

                int seqNum = dataPacket.getSeqNum();
                ackLog.println(seqNum);

                if (seqNum != packetsAcked) {
                    int offset = Math.floorMod((seqNum - packetsAcked), 32);

                    lock.lock();
                        packetsAcked = packetsAcked + offset;
                    lock.unlock();

                    lock.lock();
                        if (packetsAcked == packetsSent) stopTimer();
                        else startTimer();
                    lock.unlock();
                }
            }
            tEnd = System.currentTimeMillis();

            timeLog.println( tEnd - tStart );
        }
    }
}
